### PR TITLE
Moved include_gps and include_bipm out of Observatory objects

### LINKED ISSUE #1691: Clock corrections applied to TZRMJD even if TZRSITE is BAT
Dear all,

I am in the field of very-high energy pulsars with imaging Cherenkov telescopes such as MAGIC and CTA. Since we typically have to integrate tens of hours of observations (scattered across many months) to detect the pulsed gamma rays and build a TOA, we are forced to apply the barycenter corrections to each single _event_ we record, with most of them being just background cosmic rays, and obtain the pulse profile from already barycentered data. I find convenient for me to produce par files where TZRSITE is set to "@" or "bat", so that I can manipulate the phase=0 definition in a relatively easy way by changing TZRMJD, and adapt it to various different references present in literature.

I think I've found an inconsistency between the behaviour of PINT and tempo2 in the specific case in which TZRSITE is set to "@" and the clock is set to TT(BIPM****). Specifically, PINT applies the clock corrections to TZRMJD, whereas tempo2 doesn't. This causes the absolute phases calculated by tempo2 and PINT to differ by ~(-27us/F0), even when the BATs are identical down to the precision of np.longdouble. I was suggested to report it here by Paul Ray over the facebook pulsar group.

I believe I've identified the origin of the discrepancy in the definition of get_TZR_toa(self, toas) in $PINT/src/pint/models/absolute_phase.py and the respective function refphs_init(pulsar* psr, int npsr) in $TEMPO2/refphs.C .
The tempo2 function contains:
```
21             if (obs->telID[0]=='@'            /* Have barycentric arrival time */
22                     || strcasecmp(obs->telID,"bat")==0)
23             {
24                 obs->clockCorr=0;  /* therefore don't do clock corrections */
25                 obs->delayCorr=0;
26             }
```
whereas the PINT function does not have such a check. The docstring of get_TZR_toa() even mentions that "We are treating the TZRMJD as a special TOA. Note that any observatory clock corrections will be applied to this TOA, as with any other TOA. This does not affect the value of the TZRMJD parameter, however." Indeed, if I add such a check on the value of self.TZRSITE in get_TZR_toa() of absolute_phase.py, the discrepancy in the phases disappears:
```
111         if self.TZRSITE.value != '@' and self.TZRSITE.value.lower() != 'bat':
112             tz = toa.get_TOAs_array(
113                 (self.TZRMJD.quantity.jd1 - 2400000.5, self.TZRMJD.quantity.jd2),
114                 obs=self.TZRSITE.value,
115                 freqs=self.TZRFRQ.quantity,
116                 include_bipm=clkc_info["include_bipm"],
117                 include_gps=clkc_info["include_gps"],
118                 ephem=toas.ephem,
119                 planets=toas.planets,
120                 tzr=True,
121             )
122         else:
123             tz = toa.get_TOAs_array(
124                 (self.TZRMJD.quantity.jd1 - 2400000.5, self.TZRMJD.quantity.jd2),
125                 obs=self.TZRSITE.value,
126                 freqs=self.TZRFRQ.quantity,
127                 include_bipm=False,
128                 include_gps=False,
129                 ephem=toas.ephem,
130                 planets=toas.planets,
131                 tzr=True,
132             )
```

To compare the results, both PINT and tempo2 need to be run with the same .par and .tim files. Unfortunately I do not know a way to print the absolute phases in a human-readable format with tempo2. I've written a patch to the "general2" plugin of tempo2, that allows the fractional part of the phase to be printed using a "{phasefrac}" token (the integer part can already be printed with the {npulse} token). I've checked that the phases produced in this way are compatible with the ones produced by MAGIC's own tempo2 plugin (that writes to some horrible .root file format). Please find attached an archive with the patch and some example scripts.

## Summary

**Expected behaviour**: tempo2 and PINT phases and barycentered times should be the same when ran on the same .tim and .par files
**Observed behaviour**: PINT phases differ by an amount of  ~ (-27us/F0) when TZRSITE is "@" or "BAT" and CLK is "TT(BIPM2021)".

**Steps to reproduce**:

1. Download the archive below. The archive contains a fictional tim file for several possible observatiories, a sample par file derived from JBO's Crab monthly ephemerides, the patch to be applied to tempo2's general2 plugin, a bash script to execute tempo2, and a python script that runs PINT.
2. Recompile tempo2 with the patch that allows the absolute phases to be printed to the terminal (I am convinced there must be an easier way to do it, but I am not aware of it).
3. Run the runTempo.sh bash script that just executes "tempo2 -output general2 -f sample.par fake.tim -s "{bat} {phasefrac}\n" > tempo2_result.txt"
4. Run the testPint.py script that computes the phases and BATs with PINT and compares them with tempo2.
5. On my machine, the PINT and tempo2 BATs differ by 1e-12 MJD (i.e. at the precision of np.longdouble), whereas the phases differ by 8.17e-4 ~ 27us/F0. 

PINT version: 0.9.8+5.g0227e13c.dirty
tempo2 version: 2023.05.1 (commit 3610511302af0f6f62d2a9806177301516a9babf)

output of pint.print_info():
```
# Created: 2023-12-10T14:43:11.955338
# PINT_version: 0.9.8+5.g0227e13c.dirty
# User: giovanni
# Host: ThinkPad-T450s
# OS: Linux-6.5.0-13-generic-x86_64-with-glibc2.38
# Python: 3.11.6 (main, Oct  8 2023, 05:06:43) [GCC 13.2.0]
# endian: little
# numpy_version: 1.26.2
# numpy_longdouble_precision: float128
# scipy_version: 1.11.4
# astropy_version: 6.0.0
# pyerfa_version: 2.0.1.1
# jplephem_version: 2.21
# matplotlib_version: 3.8.2
# loguru_version: 0.7.2
# Python_prefix: /home/giovanni/.venv/pinttest
# PINT_file: /home/giovanni/.venv/pinttest/lib/python3.11/site-packages/pint/__init__.py
# Environment: virtualenv
# virtualenv_prefix: /home/giovanni/.venv/pinttest
```

[PINT-tempo2__TZRSITE-BAT.tar.gz](https://github.com/nanograv/PINT/files/13628184/PINT-tempo2__TZRSITE-BAT.tar.gz)

Thank you in advance for your help and patience, and sorry for the long message!

-- 
Dr. Giovanni Ceribella
Max-Planck-Institut für Physik
Boltzmannstr. 8
85748 Garching
ceribell@mpp.mpg.de

### PR DESCRIPTION
**Pull request by paulray**
This is an attempt to clean up handling of `include_gps` and `include_bipm`

Rather than having those as attributes of an `Observatory` object, they are passed as arguments to the `site.clock_correction()` method.  So instead of this:
```
site = get_observatory("ao", include_bipm=True)
clock_corr = site.clock_corrections(time)
```
the flow now looks like this:
```
site = get_observatory("ao")
clock_corr = site.clock_corrections(
        time, include_gps=True, include_bipm=True, bipm_version="BIPM2015"
    )
```
This has no effect on most users, since `get_TOAs()` still accepts the same arguments, it just passes `include_bipm` to the clock correction method rather than the observatory constructor.

In addition, the `include_gps` and `include_bipm` arguments now are interpreted as the user wanting UTC times to get the UTC->GPS correction added and TT times to get the TT(BIPM) corrections included.  Even if they are True, those corrections will *NOT* get applied to a time that is already in TDB.  This behavior should fix #1691.

The one significant thing on the user side that I've seen so far is that a bunch of observatories (mainly TeV and GW) in observatories.json had `include_bipm=False` included in their definition (making it their default). This is now lost. I argue this is a good thing since whether or not BIPM corrections are included is not really a feature of the observatory. So, users will now have to specify `include_bipm=False` if that is the behavior they want when loading TOAs from those observatories.

### CODE REVIEW

**Comment by paulray** [comment]
Fixes #1766

**Comment by paulray** [comment]
Also made a fix to the `clock_correction()` method of `TopoObs`, where if `self._clock` is False/None and `self.clock_files` is True, it would replace the clock corrections with 0.0. I believe this was wrong and it should just add 0.0 to the prior corrections (the BIPM and GPS corrections), so I changed `corr =` to `corr +=` in that line.

**Comment by abhisrkckl** [comment]
How does this interact with #1747?

Are clock GPS and BIPM corrections disabled when "CLOCK UNCORR" is given in the par file?

**Comment by dlakaplan** [comment]
> The one significant thing on the user side that I've seen so far is that a bunch of observatories (mainly TeV and GW) in observatories.json had `include_bipm=False` included in their definition (making it their default). This is now lost. I argue this is a good thing since whether or not BIPM corrections are included is not really a feature of the observatory. So, users will now have to specify `include_bipm=False` if that is the behavior they want when loading TOAs from those observatories.

Will most users know how/whether to do that?  Should the various wrappers to load spacecraft TOAs have that set explicitly now?

**Comment by paulray** [comment]
> > The one significant thing on the user side that I've seen so far is that a bunch of observatories (mainly TeV and GW) in observatories.json had `include_bipm=False` included in their definition (making it their default). This is now lost. I argue this is a good thing since whether or not BIPM corrections are included is not really a feature of the observatory. So, users will now have to specify `include_bipm=False` if that is the behavior they want when loading TOAs from those observatories.
> 
> Will most users know how/whether to do that? Should the various wrappers to load spacecraft TOAs have that set explicitly now?

Hmm. Most people will hopefully use `get_model_and_toas()` which will make the right choice based on the setting of CLOCK in your model.  Otherwise, it is up to the user. If they call `get_TOAs()` with no kwargs, they will get the defaults for `include_bipm` and `include_gps`, which are True.  Maybe there should not be defaults for those?  I just don't see how it makes sense for those settings to be part of the observatory. How does the fact that I'm using LIGO or HAWC imply that I don't want a BIPM timescale? I think we just need to educate people by emailing the PINT list and putting info in the release notes.

**Comment by dlakaplan** [comment]
> > > The one significant thing on the user side that I've seen so far is that a bunch of observatories (mainly TeV and GW) in observatories.json had `include_bipm=False` included in their definition (making it their default). This is now lost. I argue this is a good thing since whether or not BIPM corrections are included is not really a feature of the observatory. So, users will now have to specify `include_bipm=False` if that is the behavior they want when loading TOAs from those observatories.
> > 
> > 
> > Will most users know how/whether to do that? Should the various wrappers to load spacecraft TOAs have that set explicitly now?
> 
> Hmm. Most people will hopefully use `get_model_and_toas()` which will make the right choice based on the setting of CLOCK in your model. Otherwise, it is up to the user. If they call `get_TOAs()` with no kwargs, they will get the defaults for `include_bipm` and `include_gps`, which are True. Maybe there should not be defaults for those? I just don't see how it makes sense for those settings to be part of the observatory. How does the fact that I'm using LIGO or HAWC imply that I don't want a BIPM timescale? I think we just need to educate people by emailing the PINT list and putting info in the release notes.

If we remove the defaults that would make it more explicit that users need to set this.  Maybe that is a good plan.  But need to make sure the upstream uses like `get_fits_TOAs` also do the same.

**Comment by paulray** [comment]
The problem with removing the defaults is that everyone who has code that calls `get_TOAs()` without those kwargs will then get an exception. This might be aggravating, but on the other hand it forces people to think about what they really want.

I also looked at what happens when you call `get_TOAs()` with a `model=` argument. It can set `use_bipm`, `bipm_version` and `planets`, plus it will set `use_gps` to False *only* in the case that `CLK UNCORR` is in the model. 

So really BIPM is associated with the model and it makes sense to me that either its setting should be inherited from the model, or it should force the user to specify it explicitly.  On the other hand, I'm now re-thinking the GPS correction.  That really is a function of the observatory clock.  If the observatory clock corrections put you on UTC(GPS), as is the most common case, then you do want to apply the UTC(GPS) to UTC corrections.  Since this is site dependent, probably I should **not** have taken it out of the Observatory class.  The difference is that you should NOT use different settings for BIPM between different sites if you have multi-observatory datasets, but you SHOULD use different settings for GPS for each observatory if some use GPS clocks and some don't.

Do people agree that I should revert my changes to `use_gps`, but not `use_bipm`?

**Comment by paulray** [comment]
In thinking about `use_gps` more, I changed its name to `apply_gps2utc` since that better characterizes what it does. It really is a feature of the observatory clock, and just part of the necessary clock corrections to go from raw observatory times to UTC (along with the observatory clock file like pks2gps.clk).  So, I think it should not be a kwarg to so many functions like `get_TOAs()`.  If you really want to use an observatory and NOT apply the correction from GPS to UTC that the observatory normally thinks should be done, you could instantiate the site yourself as `site = get_observatory('gbt',apply_gps2utc=False)`.  However, this does NOT overwrite the registry so if you later call `get_observatory('gbt')` you will get the normal version with `apply_gps2utc` False.

I'd like to hear opinions but I don't think we need to make it so easy to turn off the gps2utc corrections in normal code.  That setting should be part of observatories.json and that's basically it. Thoughts?

**Comment by paulray** [comment]
Ok, question to consider. How easy should it be to not apply the UTC(GPS)->UTC corrections?
Is it sufficient to just need to modify `observatories.json` to turn it off for any observatory where you don't want it?

Or, if you want to be able to do it programmatically, what is acceptable? It is very painful to make sure that every call to `get_observatory()` has the kwarg to disable it.  But right now when you call `get_observatory('gbt',apply_gps2utc=False)` it makes a deepcopy of observatory object and then modifies that. This does NOT affect the registry and thus that setting will not "stick" so if you call `get_observatory()` again without the kwarg, the setting will disappear.  It seems to me OK to have a programmer call `get_observatory('gbt',apply_gps2utc=False,overwrite=True)` once and have that update the registry. Then future calls with no kwargs will get the observatory object with `apply_gps2utc=False`.  But this should be a pretty rare use case since normally you would want this correction for any observatory where it is correct to apply it.

**Comment by paulray** [comment]
Ah, the deepcopy was added in #1711. I am going to revert that behavior in this PR.

File: CHANGELOG-unreleased.md [hunk 1]
```diff
@@ -16,6 +16,10 @@ the released changes.
 - Type hinting for most of the `pint.utils` module
 - `funcParameter`s are no longer listed in the `pintk` interface.
 - Updated location of CCERA
+- Removed `include_bipm`, and `bipm_version` from the Observatory class. Now they are passed as arguments to `site.clock_correction()`
+- Renamed `include_gps` to `apply_gps2utc` in the Observatory class
+- Removed `apply_gps2utc` from `TOAs.clock_corr_info` since it can be different for different observatories. It is not a global setting.
+- The following observatories no longer have a default of include_bipm=False: magic, lst, virgo, lho, llo, geo600, kagra, hess, hawc
```
Comment (abhisrkckl) [inline CHANGELOG-unreleased.md:22]: `include_bipm=False` in backticks

File: src/pint/data/runtime/observatories.json [hunk 2]
```diff
@@ -1520,8 +1530,6 @@
         "origin": "An Imaging Atmospheric Cherenkov Telescope.\nThese coordinates are provided (from geodetic coordinates : lat = 23°16'18''S, lon = 16°30'00''E at 1800m asl) by the Maxime Regeard and Arache Djannati-Atai on behalf of the H.E.S.S. Collabotation, 2023 January 18."
     },
     "hawc": {
-        "include_bipm": false,
-        "include_gps": true,
```
Comment (abhisrkckl) [inline src/pint/data/runtime/observatories.json:1524]: Isn't it better to set apply_gps2utc as false explicitly?

I am assuming that's the default behaviour.

File: src/pint/observatory/__init__.py [hunk 3]
```diff
@@ -59,6 +59,7 @@
 
 # The default BIPM to use if not explicitly specified
 # FIXME: this should be auto-detected by checking the index file to see what's available
+# ALSO: Why is the a feature of the Observatory class?  Seems like the wrong place for it.
```
Comment (abhisrkckl) [inline src/pint/observatory/__init__.py:62]: May be we can move this to the `pint` namespace or `pint.toa` namespace?

File: src/pint/observatory/__init__.py [hunk 4]
```diff
@@ -272,12 +295,19 @@ def get(cls, name):
...
-            return cls._registry[cls._alias_map[name]]
+            site = cls._registry[name]
+        elif name in cls._alias_map:
+            site = cls._registry[cls._alias_map[name]]
+        else:
+            site = None
+        if site is not None:
+            if overwrite and apply_gps2utc is not None:
+                # This will modify the Observatory object in the registry, so it will "stick" until overwritten
+                site.apply_gps2utc = apply_gps2utc
```
Comment (abhisrkckl) [inline src/pint/observatory/__init__.py:309]: Can we have a warning here telling the user to be extra careful?

File: src/pint/observatory/__init__.py [hunk 5]
```diff
@@ -340,7 +371,13 @@ def timescale(self):
         astropy.time.Time()."""
         raise NotImplementedError
 
-    def clock_corrections(self, t, limits="warn"):
+    def clock_corrections(
+        self,
+        t,
```
Comment (abhisrkckl) [inline src/pint/observatory/__init__.py:376]: What is the type of t? Please add a type hint or mention it in the docstring.

File: src/pint/observatory/__init__.py [hunk 6]
```diff
@@ -469,30 +504,24 @@ def get_observatory(
...
-            site.bipm_version = bipm_version
-
-        return site
+    if apply_gps2utc is None:
+        site = Observatory.get(name)
+    else:
+        # If we are overriding the gps2utc setting then  site and update apply_gps2utc
+        # Otherwise, will just use default setting for this site.
+        log.debug(f"Observatory.get(apply_gps2utc={apply_gps2utc})")
+        site = Observatory.get(name, apply_gps2utc=apply_gps2utc, overwrite=True)
```
Comment (abhisrkckl) [inline src/pint/observatory/__init__.py:522]: A warning here?

File: src/pint/observatory/__init__.py [hunk 7]
```diff
@@ -771,7 +800,9 @@ def update_clock_files(bipm_versions=None):
     # FIXME: what to do about GPS and BIPM files?
 
     if bipm_versions is not None:
-        o = get_observatory("arecibo")
+        o = get_observatory(
+            "arecibo"
+        )  # Why arecibo? Would be better to use an observatory with no clock file of its own, right?
```
Comment (abhisrkckl) [inline src/pint/observatory/__init__.py:805]: I agree.

File: src/pint/observatory/special_locations.py [hunk 8]
```diff
@@ -86,20 +72,19 @@ def __init__(
...
     """Observatory-derived class for the solar system barycenter.
 
-    Time scale is assumed to be tdb."""
+    Time scale is assumed to be TDB, so no GPS or BIPM corrections should be applied."""
 
     def __init__(
         self,
         name,
         aliases=None,
+        apply_gps2utc=False,
```
Comment (abhisrkckl) [inline src/pint/observatory/special_locations.py:81]: Should this even be an option?

Comment (abhisrkckl) [reply src/pint/observatory/special_locations.py:81]: Is there any scenario where an SSB TOA will have a GPS correction?

File: src/pint/observatory/topo_obs.py [hunk 9]
```diff
@@ -335,7 +313,13 @@ def _load_clock_corrections(self):
                 )
             )
 
-    def clock_corrections(self, t, limits="warn"):
+    def clock_corrections(
+        self,
+        t,
```
Comment (abhisrkckl) [inline src/pint/observatory/topo_obs.py:318]: type of t?

File: CHANGELOG-unreleased.md (hunk 1 above)
Comment (paulray) [reply CHANGELOG-unreleased.md:22]: Done

File: src/pint/data/runtime/observatories.json (hunk 2 above)
Comment (paulray) [reply src/pint/data/runtime/observatories.json:1524]: The default for `include_bipm` is true (I didn't change that).
However, I agree that having it set explicitly for each observatory is good.
I've done that for all the observatories where it was obvious (e.g. the clock file was XXX2gps.clk).
The correct choice for other observatories will need to come from people with specific knowledge of how that observatory does time stamps.

File: src/pint/observatory/special_locations.py (hunk 8 above)
Comment (paulray) [reply src/pint/observatory/special_locations.py:81]: Good point. Removed.

File: src/pint/observatory/topo_obs.py (hunk 9 above)
Comment (paulray) [reply src/pint/observatory/topo_obs.py:318]: it is an astropy `Time`. It says in the docstring.

File: src/pint/observatory/__init__.py (hunk 3 above)
Comment (paulray) [reply src/pint/observatory/__init__.py:62]: I looked at this and in the end decided it was better where it was. The lowest level clock correction calculator is in the Observatory class. So I left it and deleted that comment.

What do you think about making the default auto-detect the most recent BIPM available?  I could add that.

File: src/pint/observatory/__init__.py (hunk 4 above)
Comment (paulray) [reply src/pint/observatory/__init__.py:309]: Added.

File: src/pint/observatory/__init__.py (hunk 5 above)
Comment (paulray) [reply src/pint/observatory/__init__.py:376]: Updated doctoring

File: src/pint/observatory/__init__.py (hunk 6 above)
Comment (paulray) [reply src/pint/observatory/__init__.py:522]: I added one in `Observatory.get` so no need for one here.

File: src/pint/observatory/__init__.py (hunk 3 above)
Comment (abhisrkckl) [reply src/pint/observatory/__init__.py:62]: > What do you think about making the default auto-detect the most recent BIPM available?

That makes sense.

**Comment by paulray** [comment]
This is ready to merge. @dlakaplan or @scottransom do you want to give it a look and merge? I addressed @abhisrkckl's comments.

**Comment by abhisrkckl** [comment]
I'll merge this if it is ready to go.

**Comment by paulray** [comment]
> I'll merge this if it is ready to go.

It would be great to merge soon. I just had to fix several conflicts caused by adding type hints.

**Comment by abhisrkckl** [comment]
Are you going to push more commits? If not, I'll merge this once the tests pass.

**Pull request merged**